rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+*   create models under `strategy.scope` so variables will be assigned to parameter servers.
+*   likewise, create a Keras metric object under `strategy.scope`. We expect the metric variables to be stored on parameter servers. Each worker, within their `replica_fn`, updates the metric states.
+*   use `strategy.schedule` to schedule the `replica_fn` on one remote worker. This `replica_fn` should take an iterator and perform forward and backward computation. This `strategy.schedule` will only schedule this `replica_fn` and returns one or several `Future` objects immediately.  
+*   use `strategy.local_results` to get concrete values of results returned by `strategy.schedule`. This may be a blocking call if the result is not yet ready. With any failure that cannot be handled will be ignored and as a result some of the results may be `None`.
+*   call `strategy.join` to wait until all scheduled functions are executed.
+
+```python
+# Connect to remote servers with a user-provided `ClusterResolver` object.
+strategy = ParameterServerStrategyV2(cluster_resolver)
+
+dataset_fn = # a function that returns a dataset
+
+# Clone the dataset on all workers, shuffled with different seeds.
+distributed_dataset = strategy.experimental_distribute_datasets_from_function(
+      dataset_fn)
+
+with strategy.scope():
+  # Create variables on parameter servers in a round-robin fashion.
+  model = create_model()
+  optimizer = tf.keras.optimizers.Adam()
+  accuracy = tf.keras.metrics.CategoricalAccuracy(name="train_accuracy")
+
+  @tf.function
+  def replica_fn(iterator):
+    x, y = next(iterator)
+    with tf.GradientTape() as tape:
+      predictions = model(x, table, training=True)
+      loss = compute_loss(y, predictions)
+    gradients = tape.gradient(loss, model.trainable_variables)
+    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+    accuracy.update_state(y, predictions)
+    return loss
+
+  for _ in range(num_epoches):
+    distributed_iter = iter(distributed_dataset)
+    for i in range(steps_per_epoch):
+      # strategy.schedule pushes a closure in the scheduling queue and
+      # returns a list of future objects immediately.
+      loss = strategy.schedule(replica_fn,
+                               args=(distributed_iter,))
+    strategy.join()
+    model.save()  # save checkpoint/summary...
+    print ("Loss = %f, accuracy = %f" % (
+        strategy.local_results(loss), accuracy.result()))
+```
+
+
+##### Alternative training loop: fully async
+
+Another option from calling `join` every epoch, users can choose to schedule all steps and then asynchronously print metric values. This option doesn’t require any synchronization in epoch boundaries.
+
+
+```python
+# … omitted
+with strategy.scope():
+  # … omitted
+  for _ in range(total_steps)):
+    strategy.schedule(step_fn, args=(iterators,))
```

Comment:
I don't fully understand the delta here-- what is iterators? Can you omit less?

Reply:
Sorry the name is confusing, I renamed (in my pending commit) it to `distributed_iter`.